### PR TITLE
Numark DJ2GO Touch mapping

### PR DESCRIPTION
**Pull request by tandy-1000**
I wrote a components.js mapping with help from @Holzhaus , @Swiftb0y , and others on the Zulip. (https://github.com/mixxxdj/mixxx/pull/4108)

The mapping is complete, and I have tried to write a manual entry for this based off of the Roland DJ-505 entry.

### CODE REVIEW

File: source/hardware/controllers/numark_dj2go2_touch.rst [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+-  `Manufacturer’s product page <https://www.numark.com/product/dj2go2-touch>`__
+-  `Serato DJ Hardware Page <https://serato.com/dj/hardware/numark-dj2go2-touch>`__
+-  `Mapping Forum Thread <https://mixxx.discourse.group/t/a-complete-numark-dj2go2-touch-mapping/22616>`__
+
+.. versionadded:: 2.3.0
+
+Controller Mapping
+------------------
+
+Everything in this mapping is standard except from the Level knobs, which have been remapped to a filter effect.
```
Comment (Swiftb0y) [inline source/hardware/controllers/numark_dj2go2_touch.rst:21]: This is not correct, you implemented the custom preview seek functionality on the browse knob.

File: source/hardware/controllers/numark_dj2go2_touch.rst [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+   tandy1000 <https://github.com/tandy-1000>
+
+The Numark DJ2GO2 Touch is a USB :term:`MIDI` controller with an integrated audio interface.
+It is the successor to the DJ2GO2, and features touch capable wheels.
+
+-  `Manufacturer’s product page <https://www.numark.com/product/dj2go2-touch>`__
+-  `Serato DJ Hardware Page <https://serato.com/dj/hardware/numark-dj2go2-touch>`__
+-  `Mapping Forum Thread <https://mixxx.discourse.group/t/a-complete-numark-dj2go2-touch-mapping/22616>`__
+
+.. versionadded:: 2.3.0
```
Comment (Swiftb0y) [inline source/hardware/controllers/numark_dj2go2_touch.rst:16]:
```suggestion
.. versionadded:: 2.3.1
```
2.3.0 is already released, the soonest the mapping could go into mixxx is with the 2.3.1 release.

**Review by Swiftb0y — COMMENTED** [review]
I don't know what our policy on controller manual pages is. I've never reviewed or written one so I'd prefer if someone else can take care of reviewing.

File: source/hardware/controllers/numark_dj2go2_touch.rst (hunk 1 above)
Comment (tandy-1000) [reply source/hardware/controllers/numark_dj2go2_touch.rst]: Oops, just added that in

File: source/hardware/controllers/numark_dj2go2_touch.rst [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+-  `Mapping Forum Thread <https://mixxx.discourse.group/t/a-complete-numark-dj2go2-touch-mapping/22616>`__
+
+.. versionadded:: 2.3.1
+
+Controller Mapping
+------------------
+
+Everything in this mapping is standard except:
+ - the Level knobs have been remapped to a filter effect,
+ - the browse encoder can be used to preview tracks, with an added seek functionality.
```
Comment (Swiftb0y) [inline source/hardware/controllers/numark_dj2go2_touch.rst:23]: Please be explicit. This is not a changelog, its a manual entry. Telling the user what features exist doesn't make much sense if they have no Idea how to access it. Describe how exactly this feature can be used on the controller.

Comment (tandy-1000) [reply source/hardware/controllers/numark_dj2go2_touch.rst]: good point, I just committed an improvement.

Comment (Swiftb0y) [reply source/hardware/controllers/numark_dj2go2_touch.rst]: Thanks, much clearer now.

File: source/hardware/controllers/numark_dj2go2_touch.rst [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+.. versionadded:: 2.3.1
+
+Controller Mapping
+------------------
+
+Everything in this mapping is standard except:
+ - the Level knobs have been remapped to a filter effect,
+ - the browse encoder has two added features:
+    - it can be used to preview tracks when clicking on a track in a tracklist,
```
Comment (Be-ing) [inline source/hardware/controllers/numark_dj2go2_touch.rst:24]: The formatting looks weird here:
image

File: source/hardware/controllers/numark_dj2go2_touch.rst [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+-  `Manufacturer’s product page <https://www.numark.com/product/dj2go2-touch>`__
+-  `Serato DJ Hardware Page <https://serato.com/dj/hardware/numark-dj2go2-touch>`__
+-  `Mapping Forum Thread <https://mixxx.discourse.group/t/a-complete-numark-dj2go2-touch-mapping/22616>`__
+
+.. versionadded:: 2.3.1
+
+Controller Mapping
+------------------
+
+Everything in this mapping is standard except:
```
Comment (Be-ing) [inline source/hardware/controllers/numark_dj2go2_touch.rst:21]:
```suggestion
Everything in this mapping works as labeled on the controller except:
```

File: source/hardware/controllers/numark_dj2go2_touch.rst [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ - the browse encoder has two added features:
+    - it can be used to preview tracks when clicking on a track in a tracklist,
+    - scrolling the encoder allows seeking the previewed track,
+    - clicking when a track is being previewed will stop the preview, allowing you to scroll the tracklist again.
+
+
+Known Issues
+------------
+
+- No known issues.
```
Comment (Be-ing) [inline source/hardware/controllers/numark_dj2go2_touch.rst:32]: This can be removed.

File: source/hardware/controllers/numark_dj2go2_touch.rst [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+===================
+
+.. sectionauthor::
+   tandy1000 <https://github.com/tandy-1000>
+
+The Numark DJ2GO2 Touch is a USB :term:`MIDI` controller with an integrated audio interface.
+It is the successor to the DJ2GO2, and features touch capable wheels.
+
+-  `Manufacturer’s product page <https://www.numark.com/product/dj2go2-touch>`__
+-  `Serato DJ Hardware Page <https://serato.com/dj/hardware/numark-dj2go2-touch>`__
```
Comment (Be-ing) [inline source/hardware/controllers/numark_dj2go2_touch.rst:13]: Please delete this. Numark's page is good to link to, but Serato's page doesn't help anyone use Mixxx.

**Comment by Be-ing** [comment]
Please add a brief description how the pads work.

**Comment by Be-ing** [comment]
This looks like a nice little controller. Maybe in the future when we get Mixxx working well on small screens, I'll get one to use with my PinePhone :nerd_face:

File: source/hardware/controllers/numark_dj2go2_touch.rst (hunk 4 above)
Comment (tandy-1000) [reply source/hardware/controllers/numark_dj2go2_touch.rst]: I made some changes that should fix this.

**Comment by tandy-1000** [comment]
I will add a pad guide next.

That's a great idea haha, I'm not so sure as to whether the pinephone could handle Mixxx though, I will definitely give it a go on mine at some point :)

File: source/hardware/controllers/numark_dj2go2_touch.rst (hunk 5 above)
Comment (Holzhaus) [reply source/hardware/controllers/numark_dj2go2_touch.rst:20]: Could you please add a list of controls like Roland DJ-505 mapping does?

Some users may never use the controller with another DJ software. And "everything is mapped to standard" is ambiguous. For example, what does the Sync button do? Does it do a one-shot beatsync? Or does it toggle sync lock?

**Review by Holzhaus — CHANGES_REQUESTED** [review]
Thanks.

Could you please add a list of all controls like Roland DJ-505 mapping does?

Some users may never use the controller with another DJ software. And "everything works as labeled on the controller" is ambiguous. For example, what does the Sync button do? Does it do a one-shot beatsync? Or does it toggle sync lock? For this reason, please document each button individually.

File: source/hardware/controllers/numark_dj2go2_touch.rst [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+------------------
+
+Everything in this mapping is standard except:
+
+* the Level knobs have been remapped to a filter effect,
+* the browse encoder has two added features:
+
+  * it can be used to preview tracks when clicking on a track in a tracklist, scrolling the encoder allows seeking the previewed track,
+  
+  * clicking when a track is being previewed will stop the preview, allowing you to scroll the tracklist again.
```
Comment (Holzhaus) [inline source/hardware/controllers/numark_dj2go2_touch.rst:27]: Please add the browse knob to the table with the "Mixer" section controls (maybe rename it to "Mixer/Library"), then we can remove this section entirely,

**Comment by tandy-1000** [comment]
I've improved the manual to include more things like the DJ-505 manual entry.

File: source/hardware/controllers/numark_dj2go2_touch.rst [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+===============================================================  ======================================================================================================================
+Control                                                          Function
+===============================================================  ======================================================================================================================
+:hwlabel:`CUES` mode                                             Set cue points, and trigger them by clicking.
+:hwlabel:`AUTO` mode                                             Enable / disable beat loops, with pads 1, 2, 3, and 4 creating 1, 4, 8, and 16 bar loops respectively.
+:hwlabel:`MANUAL` mode                                           Set a custom loop: pad 1 sets the loop in point, pad 2 sets the loop out point, and pad 3 enables / disables the loop.
+:hwlabel:`SAMPLER` mode                                          Load / trigger sample tracks.
+===============================================================  ======================================================================================================================
```
Comment (Holzhaus) [inline source/hardware/controllers/numark_dj2go2_touch.rst:59]:
```suggestion
===============================================================  ======================================================================================================================
Pad Mode                                                         Function
===============================================================  ======================================================================================================================
:hwlabel:`CUES`                                                  Set cue points, and trigger them by clicking.
:hwlabel:`AUTO`                                                  Enable / disable beat loops, with pads 1, 2, 3, and 4 creating 1, 4, 8, and 16 bar loops respectively.
:hwlabel:`MANUAL`                                                Set a custom loop: pad 1 sets the loop in point, pad 2 sets the loop out point, and pad 3 enables / disables the loop.
:hwlabel:`SAMPLER`                                               Load / trigger sample tracks.
===============================================================  ======================================================================================================================
```

File: source/hardware/controllers/numark_dj2go2_touch.rst [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+Mixer
+~~~~~
+
+===============================================================  ==================================
+Control                                                          Function
+===============================================================  ==================================
+:hwlabel:`MASTER LEVEL`                                          Adjusts the master output level.
+:hwlabel:`CUE LEVEL`                                             Adjusts the cue output level.
+Cross fader                                                      Fades between left and right deck.
```
Comment (Holzhaus) [inline source/hardware/controllers/numark_dj2go2_touch.rst:69]:
```suggestion
Crossfader                                                       Fades between left and right deck.
```

**Review by Holzhaus — COMMENTED** [review]
Thanks, looks good already!

File: source/hardware/controllers/numark_dj2go2_touch.rst [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+===============================================================  ======================================================
+Control                                                          Function
+===============================================================  ======================================================
+Jog Wheel (top surface)                                          Perform scratch operation.
+Jog Wheel (outer edge)                                           Rotate to lower/raise playback speed and pitch.
+:hwlabel:`SYNC`                                                  Match tempo and phase of other deck.
+:hwlabel:`CUE`                                                   Specifies, plays or recalls temporary cue point.
+:hwlabel:`PLAY/PAUSE`                                            Plays or pause the song.
+:hwlabel:`HEADPHONES`                                            Switches audio output between headphones and speakers.
```
Comment (Be-ing) [inline source/hardware/controllers/numark_dj2go2_touch.rst:31]:
```suggestion
:hwlabel:`HEADPHONES`                                            Routes deck to headphones output
```

File: source/hardware/controllers/numark_dj2go2_touch.rst [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+~~~~
+
+The pad mode is changed using the :hwlabel:`PAD MODE` button.
+
+===============================================================  ======================================================================================================================
+Pad Mode                                                         Function
+===============================================================  ======================================================================================================================
+:hwlabel:`CUES`                                                  Set cue points, and trigger them by clicking.
+:hwlabel:`AUTO`                                                  Enable / disable beat loops, with pads 1, 2, 3, and 4 creating 1, 4, 8, and 16 bar loops respectively.
+:hwlabel:`MANUAL`                                                Set a custom loop: pad 1 sets the loop in point, pad 2 sets the loop out point, and pad 3 enables / disables the loop.
```
Comment (Be-ing) [inline source/hardware/controllers/numark_dj2go2_touch.rst:47]: You could use the 4th pad for reloop_andstop

File: source/hardware/controllers/numark_dj2go2_touch.rst [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+:hwlabel:`SAMPLER`                                               Load / trigger sample tracks.
+===============================================================  ======================================================================================================================
+
+Mixer / Library
+~~~~~~~~~~~~~~~
+
+===============================================================  ==================================
+Control                                                          Function
+===============================================================  ==================================
+Browse encoder                                                   Preview tracks when clicking on a track in a tracklist, scrolling the encoder allows seeking the previewed track. Clicking when a track is being previewed will stop the preview, allowing you to scroll the tracklist again.
```
Comment (Be-ing) [inline source/hardware/controllers/numark_dj2go2_touch.rst:57]:
```suggestion
Browse encoder                                                   Scroll in the track list. Preview tracks when clicking on a track in a tracklist, scrolling the encoder allows seeking the previewed track. Clicking when a track is being previewed will stop the preview, allowing you to scroll the tracklist again.
```

**Comment by tandy-1000** [comment]
Forgot to say; can someone review these fixes?

**Review by Swiftb0y — APPROVED** [review]
LGTM tbh.
Does somebody else want to take another look? I'm not super familiar with our policies on manual entries.

**Comment by Holzhaus** [comment]
> I'm not super familiar with our policies on manual entries.

- Make sure the markup is correct (`:hwlabel:`, `:file:`, etc.) and no links are broken in the new file
- ensure that `versionadded::`/`versionchanged::` is used
- Google a photo of the controller and check the labels mentioned here are correct
- check that the controller behavior makes sense (does not deviate from labels for no reason). This is technically not an issue with the documentation but whatever ;-)
- check that all controls are explained in a table so you can easily look up what a certain button does
- unmapped/non-working controls should be mentioned in a known issues section at the bottom
- optional but appreciated: google the hardware manual PDF and check if the manual contains a schematic view of the controller as a vector graphic. If so, we could open the PDF with Inkscape, save the graphic as SVG and add it to this manual page

**Comment by tandy-1000** [comment]
I'll try to go through that list soon.

**Comment by tandy-1000** [comment]
Is there a deadline I should meet to get this in with 2.3.1?

**Comment by tandy-1000** [comment]
Just added in an image of the controller, that should be everything on the list now.

**Comment by Swiftb0y** [comment]
Thank you very much for this addition. Please make sure to also set up pre-commit for the manual repository so the spell-check and SVG optimizer runs.

**Comment by tandy-1000** [comment]
Oh sorry, I'm on a different machine and forgot to do that.

**Comment by tandy-1000** [comment]
I set it up and there were no changes, the SVG was already optimised :).

**Comment by Swiftb0y** [comment]
CI complains that it isn't (which makes sense since it would be very unusual for an optimized SVG to be an entire MB big): https://github.com/mixxxdj/manual/pull/411/checks?check_run_id=3184464365

Pre-commit only runs on files being committed. Since you already committed the the unoptimized SVG, pre-commit didn't run on it.
You can run pre-commit manually using `pre-commit run --files [FILES...]`.

**Comment by Holzhaus** [comment]
Thanks! One last addition: could you add a section regarding audio setup like this? https://deploy-preview-411--mixxx-manual.netlify.app/hardware/controllers/roland_dj_505.html#audio-setup

**Comment by tandy-1000** [comment]
Done.

**Review by Holzhaus — APPROVED** [review]

**Review by Swiftb0y — APPROVED** [review]
Thanks. I'll go ahead and merge.

**Pull request merged**